File: frontend/src/app/core/utils/documento-validator.ts
/** Validação alinhada a DocumentoValidator.cs (RN-Gerais 14/15). */
export function validarCpf(cpf: string): boolean {
  const d = (cpf || '').replace(/\D/g, '');
  if (d.length !== 11 || /^(\d)\1+$/.test(d)) return false;
  let s = 0;
  for (let i = 0; i < 9; i++) s += +d[i] * (10 - i);
  let r = s % 11;
  const d1 = r < 2 ? 0 : 11 - r;
  s = 0;
  for (let i = 0; i < 10; i++) s += +d[i] * (11 - i);
  r = s % 11;
  const d2 = r < 2 ? 0 : 11 - r;
  return d.endsWith(`${d1}${d2}`);
}

export function validarCnpjNumerico(cnpj: string): boolean {
  const d = (cnpj || '').replace(/\D/g, '');
  if (d.length !== 14 || /^(\d)\1+$/.test(d)) return false;
  const w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  const w2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  let s = 0;
  for (let i = 0; i < 12; i++) s += +d[i] * w1[i];
  let r = s % 11;
  const d1 = r < 2 ? 0 : 11 - r;
  s = 0;
  for (let i = 0; i < 12; i++) s += +d[i] * w2[i];
  s += d1 * w2[12];
  r = s % 11;
  const d2 = r < 2 ? 0 : 11 - r;
  return d.endsWith(`${d1}${d2}`);
}

function valorCnpjAlfa(c: string): number {
  const u = c.toUpperCase();
  return u >= '0' && u <= '9' ? u.charCodeAt(0) - 48 : u.charCodeAt(0) - 55;
}

export function validarCnpjAlfanumerico(cnpj: string): boolean {
  const raw = (cnpj || '').replace(/[^0-9A-Za-z]/g, '').toUpperCase();
  if (raw.length !== 14) return false;
  const w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  const w2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  let s = 0;
  for (let i = 0; i < 12; i++) s += valorCnpjAlfa(raw[i]) * w1[i];
  let r = s % 11;
  const dv1 = r < 2 ? 0 : 11 - r;
  s = 0;
  for (let i = 0; i < 12; i++) s += valorCnpjAlfa(raw[i]) * w2[i];
  s += dv1 * w2[12];
  r = s % 11;
  const dv2 = r < 2 ? 0 : 11 - r;
  return raw[12] === String(dv1) && raw[13] === String(dv2);
}

export function validarCnpj(cnpj: string): boolean {
  const raw = (cnpj || '').replace(/[^0-9A-Za-z]/g, '');
  if (/^\d+$/.test(raw)) return validarCnpjNumerico(raw);
  return validarCnpjAlfanumerico(raw);
}

export function validarCpfCnpj(valor: string): boolean {
  const raw = (valor || '').replace(/[^0-9A-Za-z]/g, '');
  if (raw.length <= 11) return validarCpf(raw.padStart(11, '0').slice(-11));
  return validarCnpj(valor);
}

export function mascaraCpf(v: string): string {
  const d = v.replace(/\D/g, '').slice(0, 11);
  if (d.length <= 3) return d;
  if (d.length <= 6) return `${d.slice(0, 3)}.${d.slice(3)}`;
  if (d.length <= 9) return `${d.slice(0, 3)}.${d.slice(3, 6)}.${d.slice(6)}`;
  return `${d.slice(0, 3)}.${d.slice(3, 6)}.${d.slice(6, 9)}-${d.slice(9)}`;
}

export function mascaraCnpj(v: string): string {
  const d = v.replace(/[^0-9A-Za-z]/g, '').toUpperCase().slice(0, 14);
  if (d.length <= 2) return d;
  if (d.length <= 5) return `${d.slice(0, 2)}.${d.slice(2)}`;
  if (d.length <= 8) return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5)}`;
  if (d.length <= 12) return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5, 8)}/${d.slice(8)}`;
  return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5, 8)}/${d.slice(8, 12)}-${d.slice(12)}`;
}

export function mascaraCep(v: string): string {
  const d = v.replace(/\D/g, '').slice(0, 8);
  if (d.length <= 5) return d;
  return `${d.slice(0, 5)}-${d.slice(5)}`;
}

export function somenteDigitos(v: string): string {
  return (v || '').replace(/\D/g, '');
}

File: WebApi/AppCondominium.Domain/Validators/DocumentoValidator.cs
using System.Linq;

namespace AppCondominium.Domain.Validators;

public static class DocumentoValidator
{
    public static bool ValidarCpf(string cpf)
    {
        if (string.IsNullOrWhiteSpace(cpf)) return false;
        cpf = new string(cpf.Where(char.IsDigit).ToArray());
        if (cpf.Length != 11) return false;
        if (cpf.Distinct().Count() == 1) return false;

        int[] m1 = { 10, 9, 8, 7, 6, 5, 4, 3, 2 };
        int[] m2 = { 11, 10, 9, 8, 7, 6, 5, 4, 3, 2 };
        var temp = cpf[..9];
        int soma = 0;
        for (int i = 0; i < 9; i++) soma += (temp[i] - '0') * m1[i];
        int d1 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        temp += d1;
        soma = 0;
        for (int i = 0; i < 10; i++) soma += (temp[i] - '0') * m2[i];
        int d2 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        return cpf.EndsWith($"{d1}{d2}");
    }

    public static bool ValidarCnpjNumerico(string cnpj)
    {
        if (string.IsNullOrWhiteSpace(cnpj)) return false;
        cnpj = new string(cnpj.Where(char.IsDigit).ToArray());
        if (cnpj.Length != 14) return false;
        if (cnpj.Distinct().Count() == 1) return false;

        int[] m1 = { 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };
        int[] m2 = { 6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };
        var temp = cnpj[..12];
        int soma = 0;
        for (int i = 0; i < 12; i++) soma += (temp[i] - '0') * m1[i];
        int d1 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        temp += d1;
        soma = 0;
        for (int i = 0; i < 13; i++) soma += (temp[i] - '0') * m2[i];
        int d2 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        return cnpj.EndsWith($"{d1}{d2}");
    }

    public static bool ValidarCnpjAlfanumerico(string cnpj)
    {
        if (string.IsNullOrWhiteSpace(cnpj)) return false;
        Span<char> buffer = stackalloc char[14];
        int pos = 0;
        foreach (char c in cnpj)
        {
            if (char.IsLetterOrDigit(c))
            {
                if (pos >= 14) return false;
                buffer[pos++] = char.ToUpperInvariant(c);
            }
        }
        if (pos != 14) return false;

        int[] p1 = { 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };
        int[] p2 = { 6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };
        int soma = 0;
        for (int i = 0; i < 12; i++) soma += Valor(buffer[i]) * p1[i];
        int dv1 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        soma = 0;
        for (int i = 0; i < 12; i++) soma += Valor(buffer[i]) * p2[i];
        soma += dv1 * p2[12];
        int dv2 = soma % 11 < 2 ? 0 : 11 - soma % 11;
        return buffer[12] == (char)(dv1 + '0') && buffer[13] == (char)(dv2 + '0');
    }

    public static bool ValidarCnpj(string cnpj)
    {
        if (string.IsNullOrWhiteSpace(cnpj)) return false;
        var raw = new string(cnpj.Where(char.IsLetterOrDigit).ToArray());
        if (raw.All(char.IsDigit)) return ValidarCnpjNumerico(raw);
        return ValidarCnpjAlfanumerico(raw);
    }

    private static int Valor(char c) => char.IsDigit(c) ? c - '0' : c - 'A' + 10;
}
